### DISCUSSION TITLE
Excellent template but how do I generate the docs to my C library within the repository so that I can know how it would look like.

### DISCUSSION THREAD

**Discussion by jymchng** [friendlyanon/cmake-init#115] [Q&A]
As titled and thank you for doing this.

**Comment by friendlyanon** [comment]
The situation in general around documentation is not quite satisfying to me. See these issues: https://github.com/friendlyanon/cmake-init/issues?q=is%3Aissue+is%3Aopen+label%3Adocs

**Reply by jymchng** [reply]
Thank you for your prompt reply.

In your `CMakePresets.json`, which I copied below

```
      "name": "ci-std",
      "description": "This preset makes sure the project actually builds with at least the specified standard",
      "hidden": true,
      "cacheVariables": {
        "CMAKE_C_EXTENSIONS": "OFF",
        "CMAKE_C_STANDARD": "23",
        "CMAKE_C_STANDARD_REQUIRED": "ON"
      }
    },
    {
      "name": "flags-gcc-clang",
      "description": "These flags are supported by both GCC and Clang",
      "hidden": true,
      "cacheVariables": {
        "CMAKE_C_FLAGS": "-U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -fcf-protection=full -fstack-clash-protection -Wall -Wextra -Wpedantic -Wconversion -Wsign-conversion -Wcast-qual -Wformat=2 -Wundef -Werror=float-equal -Wshadow -Wcast-align -Wunused -Wnull-dereference -Wdouble-promotion -Wimplicit-fallthrough -Werror=strict-prototypes -Wwrite-strings",
        "CMAKE_EXE_LINKER_FLAGS": "-Wl,--allow-shlib-undefined,--as-needed,-z,noexecstack,-z,relro,-z,now,-z,nodlopen",
        "CMAKE_SHARED_LINKER_FLAGS": "-Wl,--allow-shlib-undefined,--as-needed,-z,noexecstack,-z,relro,-z,now,-z,nodlopen"
      }
    },
    {
      "name": "flags-appleclang",
      "hidden": true,
      "cacheVariables": {
        "CMAKE_C_FLAGS": "-fstack-protector-strong -Wall -Wextra -Wpedantic -Wconversion -Wsign-conversion -Wcast-qual -Wformat=2 -Wundef -Werror=float-equal -Wshadow -Wcast-align -Wunused -Wnull-dereference -Wdouble-promotion -Wimplicit-fallthrough -Werror=strict-prototypes -Wwrite-strings"
      }
    },
    {
      "name": "flags-msvc",
      "description": "Note that all the flags after /W4 are required for MSVC to conform to the language standard",
      "hidden": true,
      "cacheVariables": {
        "CMAKE_C_FLAGS": "/sdl /guard:cf /utf-8 /diagnostics:caret /w14165 /w44242 /w44254 /w34287 /w44296 /w44365 /w44388 /w44464 /w14545 /w14546 /w14547 /w14549 /w14555 /w34619 /w44774 /w44777 /w24826 /w14905 /w14906 /w14928 /W4 /permissive- /volatile:iso /Zc:inline /Zc:preprocessor",
        "CMAKE_EXE_LINKER_FLAGS": "/machine:x64 /guard:cf"
      }
    },
    {
      "name": "ci-linux",
      "inherits": ["flags-gcc-clang", "ci-std"],
      "generator": "Unix Makefiles",
      "hidden": true,
      "cacheVariables": {
        "CMAKE_BUILD_TYPE": "Release"
      }
    },
```

The `CMAKE_BUILD_TYPE` is set to `Release` in the `dev-mode` preset, why is it not set to `Debug`, for usage in say `ci-linux` preset? Or should the developer build with `CMAKE_BUILD_TYPE` with value `DEBUG` in local development and the other presets are meant for CICD pipeline, e.g GitHub Actions?

**Reply by jymchng** [reply]
I also agree that the `docs.cmake` and `docs-ci.cmake` are quite fragile.

For example, I use `cmake-format` to format my `.cmake` files and I like to UPPERCASE all my commands.

Hence, this code snippet will fail:

```
# ---- Process project() call in CMakeLists.txt ----

FILE(READ "${src}/CMakeLists.txt" content)
MESSAGE(STATUS "[INFO] `\${src}/CMakeLists.txt` = ${src}/CMakeLists.txt")

STRING(FIND "${content}" "project(" index)
IF(index EQUAL "-1")
    MESSAGE(FATAL_ERROR "Could not find \"project(\"")
ENDIF()
STRING(SUBSTRING "${content}" "${index}" -1 content)

STRING(FIND "${content}" "\n)\n" index)
IF(index EQUAL "-1")
    MESSAGE(FATAL_ERROR "Could not find \"\\n)\\n\"")
ENDIF()
STRING(SUBSTRING "${content}" 0 "${index}" content)
```

Since in my `${src}/CMakeLists.txt`, my `PROJECT(...)` would be capitalized`.

**Reply by jymchng** [reply]
Or maybe you can add in a comment like so:

```
# ---- Process project() call in CMakeLists.txt ----
# The aim is to obtain the entire 
# ```
# project (
    # PROJECT_NAME
    # ...
    # ...
# )
# ```
# call out from your `${src}/CMakeLists.txt` file
```

So that users can edit the codes in `docs-ci.json` accordingly to fit how the `project(...)` call looks like.

**Reply by friendlyanon** [reply]
> I like to UPPERCASE all my commands

Please don't do that. Use commands as they are documented. This also makes grepping for them difficult. In that case, the docs scripts breaking are absolutely a positive thing.

**Reply by jymchng** [reply]
@friendlyanon Is it a standard practice to leave the functions / calls as case insensitive?

I will heed your advice since I am only 1 week old into `Cmake`.

By the way, is it also standard practice to use the top level makefile as a command runner?

Thank you.

**Reply by friendlyanon** [reply]
> Is it a standard practice to leave the functions / calls as case insensitive?

I'm unsure what this means. If the documentation of `if()` references it as `if()`, then use it as `if()`. Same goes for commands like `FetchContent_MakeAvailable()`.

> is it also standard practice to use the top level makefile as a command runner?

Not sure what this means either. The root level `CMakeLists.txt` is the entrypoint into your build. It's by convention in the root directory for ease of use and exposure.
